gk/main.c

```diff
@@ -91,30 +628,94 @@ run_gk(struct gk_config *gk_conf)
 	/* TODO Initialize and run GK functional block. */
 
 	unsigned int i;
+	unsigned int num_lcores;
 	int ret;
+	uint8_t port_in = get_net_conf()->front.id;
+	uint16_t gk_queues[GK_MAX_NUM_LCORES];
 
-	if (!gk_conf)
-		return -1;
+	if (!gk_conf) {
+		ret = -1;
+		goto out;
+	}
+
+	num_lcores = gk_conf->lcore_end_id - gk_conf->lcore_start_id + 1;
+	RTE_ASSERT(num_lcores <= GK_MAX_NUM_LCORES);
+	for (i = 0; i < num_lcores; i++) {
+		gk_queues[i] = i + gk_conf->lcore_start_id;
+	}
+
+	ret = gatekeeper_setup_rss(port_in, gk_queues, num_lcores);
+	if (ret < 0)
+		goto out;
+
+	/* Convert RSS key. */
+	rte_convert_rss_key((uint32_t *)&default_rss_key,
+		(uint32_t *)rss_key_be, RTE_DIM(default_rss_key));
+
+	gk_conf->instances = (struct gk_instance *)rte_calloc(NULL,
+		num_lcores, sizeof(struct gk_instance), 0);
+	if (!gk_conf->instances) {
+		ret = -1;
+		goto out;
+	}
 
 	for (i = gk_conf->lcore_start_id; i <= gk_conf->lcore_end_id; i++) {
+		/* Setup the gk instance for lcore @i. */
+		ret = setup_gk_instance(i, gk_conf);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER, 
+				"gk: failed to setup gk instances for GK block at %u!\n",
+				i);
+			goto setup;
+		}
+
 		ret = rte_eal_remote_launch(gk_proc, gk_conf, i);
 		if (ret) {
 			RTE_LOG(ERR, EAL, "lcore %u failed to launch GK\n", i);
-			return ret;
+			ret = -1;
+			goto instance;
 		}
 	}
 
-	return 0;
+	goto out;
+
+instance:
+	/* GK instance at lcore @i failed to launch. */
+	rte_hash_free(gk_conf->instances[i - gk_conf->lcore_start_id].ip_flow_hash_table);
+	rte_free(gk_conf->instances[i - gk_conf->lcore_start_id].ip_flow_entry_table);
+setup:
+	/*
+	 * Failed to setup the first gk instance, needs to release
+	 * 'gk_conf->instances' and 'gk_conf'.
+	 * Otherwise, the launched gk instances need to call cleanup_gk() to release.
+	 */
+	if (i == gk_conf->lcore_start_id) {
+		rte_free(gk_conf->instances);
+		rte_free(gk_conf);
```

Comment:
Agree. However, in our case, the caller for this function is the Lua script. I wonder do we need to free ``gk_conf`` in the ``gk.lua``? Maybe simply freeing it here is simpler.